src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -130,7 +133,11 @@ Score(const size_t queryIndex, TreeType& referenceNode)
   const double minKernel = kernel.Evaluate(maxDistance);
   const double bound = maxKernel - minKernel;
 
-  if (bound <= (absError + relError * minKernel) / referenceSet.n_cols)
+  // Error tolerance of the current query point and reference node.
+  const double errorTolerance =
+      (absError + relError * minKernel) / referenceSet.n_cols;
```

Comment:
I just realized, I think that maybe this should be `referenceNode.NumDescendants()` not `referenceSet.n_cols`.  The reason is this:

For simplicity let's consider relative error only and also single-tree only.  Our goal for KDE is to ensure that, overall,

```
| K(p_q) - K'(p_q) | < \epsilon * K(p_q)
```

where `K(p_q)` is the true kernel density estimate for `q` and `K'(q)` is our estimate.  Also, `K'(p_q) = k'(p_q, p_r0) + k'(p_q, p_r1) + ... + k'(p_q, p_rn)` where `k'(q, r_i)` is our kernel function estimate between the points `p_q` and `p_ri`.

We do this by making sure that the following is true for each point:

```
| k(p_q, p_ri) - k'(p_q, p_ri) | <= \epsilon * k(p_q, p_ri)
```

(I will finish this comment, but from a different device.  That means I have to hit "add single comment" now, and I'll add more later.)

Reply:
In the single tree case, where we have a node `N_r`, we simply need to ensure that this is true for each descendant point `r_di`.  Given `k_max` and `k_min` as the maximum and minimum kernel evaluations between `p_q` and any `r_di`, we have that

```
| k(p_q, r_di) - k'(p_q, r_di) | <= | k_max - k_min |
```

for any `k'(p_q, r_di)` and actually if we are assuming that we will take `k'(p_q, r_di) = (k_max - k_min) / 2` we can get an even tighter bound:

```
| k(p_q, r_di) - k'(p_q, r_di) | <= | k_max - k_min | / 2.
```

and then, we also have that `k_min <= k(p_q, r_di)`.  So, we can plug this into both sides of our per-point bound at the bottom of our last comment to get

```
| k_max - k_min | / 2 <= \epsilon * k_min.
| k_max - k_min | <= \epsilon * k_min / 2.
```

This actually gives a very different pruning rule for relative error than what we have implemented, and while I was doing this derivation I realized that it will have a different form for absolute error.  I will try and derive that later.

Reply:
![IMG_20190820_161221](https://user-images.githubusercontent.com/1845039/63396287-85455380-c394-11e9-97ae-67ba0bd9a094.jpg)

I derived it again for absolute error, and also found that the tests all pass when I change the code.  So I think that it is correct.  I haven't yet figured out what the "combined form" of the bound is though (maybe it's easy).

In fact, as you can see, the bound for relative error came out even tighter than I thought---instead of dividing by `referenceNode.NumDescendants()`, we can multiply by 2!

Reply:
Finally, when I derived the "combined bound", I get the following:

We require that our kernel density estimate `K'(q)` satisfies

```
| K(q) - K'(q) | <= \epsilon_abs + \epsilon_rel * K(q).
```

Let's consider only a single node, in which case we require that for a reference node N_r,

```
\sum_{p_ri \in N_r} | k(q, p_ri) - k'(q, p_ri) | <= (| N_r | / | S_r |) \epsilon_abs + \epsilon_rel (\sum_{p_ri \in N_r} | k(q, p_ri) |
| N_r | | k_max(q, N_r) - k_min(q, N_r) | / 2 <= (| N_r | / | S_r |) \epsilon_abs + \epsilon_rel * | N_r | k_min(q, N_r)
| k_max(q, N_r) - k_min(q, N_r) | <= 2 \epsilon_abs / | S_r | + 2 \epsilon_rel k_min(q, N_r).
```

So, this means that the single-tree bound should be

```
bound <= (2.0 * absError / referenceSet.n_cols) + 2.0 * relError * minKernel
```

and similarly for the dual-tree bound.

So, I guess, at the end of this long journey:

(1) our bounds for KDE pruning were very loose, and we can do much better!
(2) the original error stems from the Tree-Independent Dual-Tree Algorithms paper
(3) the `referenceSet.n_cols` term is indeed part of the bound, but only for the absolute error part
(4) there is an extra factor of 2 in the bounding we can take advantage of
(5) the relative error bound term doesn't need to depend on `referenceSet.n_cols` at all

Anyway, if anything about this is unclear, just let me know.  We should make these changes, but it's up to you if we do it in this PR or another one.  But any benchmarking we do of the error reclamation should be done after these changes are made, since they'll make a huge performance difference (I think). :+1:
```

Reply:
Last comment for the night, I promise---

I ran the code with the rederived bound and it seems to work.  But there are some places in `kde_main_test.cpp` where we have to replace the `relError` in `BOOST_REQUIRE_CLOSE()` with `relError * 100` (just like in `kde_test.cpp`).